### DISCUSSION TITLE
FreeRtos stopss logger Stream

### DISCUSSION THREAD

**Discussion by brinzebuah** [modm-io/modm#1246] [Q&A] [answered]
Board: Nucleo F439ZI
I have a class which prints strings via the modm logger. As soon as i add a Mutex as an Class-Attribut I receive with hTerm only the first letter of the String, even if i not use the Mutex at all. If I initilazie the Mutex with an nullPtr I receive the whole string.
Why is this happening?
I'm a FreeRtos Noob, so i'm happy for every input.

Class "testMutex.cpp" : 

```cpp
#include <modm/board.hpp>
#include <modm/processing.hpp>

class TestMutex
{
public:
    TestMutex() = default;

    void
    hello()
    {
        modm::log::debug << "Welcome to Debug Stream example modm" << modm::endl;
    }

private:
    modm::rtos::Mutex mMutex;
    int mValue;
};
```

main:
```cpp
#include "testMutex.cpp"
#include <modm/board.hpp>

#undef MODM_LOG_LEVEL
#define MODM_LOG_LEVEL modm::log::DEBUG

int
main()
{
    Board::initialize();
    TestMutex helloMutex;
    helloMutex.hello();
}
```

**Comment by salkinium** [comment]
Probably best to start with [this example of FreeRTOS integration](https://github.com/modm-io/modm/blob/develop/examples/nucleo_g071rb/rtos/main.cpp) in modm.
Here's another example of [using FreeRTOS without a heap](https://github.com/modm-io/modm/blob/develop/examples/nucleo_f429zi/freertos_static/main.cpp).
Here's also an example of using [the FreeRTOS UART queues](https://github.com/modm-io/modm/blob/39b77479ed008b2d9851545e9dc6cbbf3afa2c0f/examples/black_pill_f401/uart_freertos/main.cpp) introduced in #1156.

Here you can [find all FreeRTOS examples](https://github.com/search?q=repo%3Amodm-io%2Fmodm+%22modm%3Afreertos%22+path%3A%2F%5Eexamples%5C%2F%2F&type=code).

**Comment by brinzebuah** [comment, marked as answer]
I Found the problem in the modm::platform::Usart3::write(uint8_t data) methode. As soon as the  UsartHal3::isTransmitRegisterEmpty() is false, the buffer gets filled. But the Interrupt, which should empty the txBuffer will never be activated, as soon as I initilazie a Mutex anywhere. 
My solution was to add a new Function emptyBuffer() to the Usart3, which does the same as the MODM_ISR(USART3) Interrupt. It enables the possibility to empty the Buffer without an ISR, and switch it into a Thread if needed. I don't really now how this ISR works and if i disablede/acknowledged all of them. Could you give me some feedback, if this is a possible solution or opens new problems? And where should I disable/acknowledge the ISR?
```void
Usart3::emptyBuffer()
{
	while(!txBuffer.isEmpty())
	{	
		if (UsartHal3::isTransmitRegisterEmpty()) {
			UsartHal3::write(txBuffer.get());
			txBuffer.pop();
		}
	}
	UsartHal3::disableInterrupt(UsartHal3::Interrupt::TxEmpty);
	UsartHal3::acknowledgeInterruptFlags(UsartHal3::InterruptFlag::OverrunError);
}
```

**Reply by salkinium** [reply]
Are you using the [FreeRTOS UART queues](https://github.com/modm-io/modm/blob/39b77479ed008b2d9851545e9dc6cbbf3afa2c0f/examples/black_pill_f401/uart_freertos/main.cpp)?
I'm not super familiar with FreeRTOS (or the UART code), I'm using NuttX at work, so I cannot tell what the potential issues are off the cuff.

Tbh, if you're looking for a C++ RTOS, modm is not going to make you happy, it was not designed with that in mind. We integrated FreeRTOS to be able to schedule "long running" tasks (historically an A* route planner) in a background thread while the main tasks are scheduled cooperatively using Protothreads or Fibers. So the peripheral drivers are all polling based, thus incompatible with RTOSes, which are event based.

**Reply by brinzebuah** [reply]
No i use the Logger, which is implementet directly with Borad::initialize().
Could the polling based thing be a reason, why Multithreading don't work? When i only use the main to initilaize Threads and start the scheduler, it only runs for 1 or 2 rounds, than i get an assertion error or end in a deadlock.
